File: src/zoho_books_cli/commands/bills.py
"""`zb bills ...` — full coverage of /bills.

Thin wrappers: CRUD + update-by-custom-field + state actions
(mark-void / mark-open) + email + payments-applied sub-app +
comments (read-only) + attachments (add / get / delete).

The collection key used by Zoho is `bills` for list and `bill` for the
single-object envelope; envelope normalization is handled by
`_shared.emit_list_paginated` / `emit_object`.
"""

from __future__ import annotations

from pathlib import Path

import typer

from zoho_books_cli import _uploads, config, output
from zoho_books_cli.client import ZohoBooksClient
from zoho_books_cli.commands import _shared
from zoho_books_cli.errors import ZohoCLIError

app = typer.Typer(
    help="Bill operations (CRUD + state + email + payments + comments + attachments).",
    no_args_is_help=True,
)
attachments_app = typer.Typer(help="Manage attachments on a bill.", no_args_is_help=True)
payments_app = typer.Typer(help="Payments applied to a bill.", no_args_is_help=True)
comments_app = typer.Typer(help="Bill history and comments (read-only).", no_args_is_help=True)
app.add_typer(attachments_app, name="attachments")
app.add_typer(payments_app, name="payments")
app.add_typer(comments_app, name="comments")

BASE = "/bills"


@app.command("list")
def list_bills(
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List bills. Returns one page plus page_context."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            BASE,
            q,
            "bills",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


@app.command("create")
def create_bill(
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Create a bill."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(BASE, json_body=json_body)
    _shared.emit_object(resp)


@app.command("get")
def get_bill(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
):
    """Get a single bill by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{bill_id}")
    _shared.emit_object(resp)


@app.command("update")
def update_bill(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a bill by ID."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{bill_id}", json_body=json_body)
    _shared.emit_object(resp)


@app.command("update-by-custom-field")
def update_bill_by_custom_field(
    key: str = typer.Option(..., "--key", help="Custom-field API name."),
    value: str = typer.Option(..., "--value", help="Custom-field value to match on."),
    body: str = typer.Option(
        ..., "--body", "-b", help="JSON body with the update fields. IDs must be strings."
    ),
    upsert: bool = typer.Option(False, "--upsert", help="Create a new bill if no match is found."),
):
    """Update a bill by a custom field's unique value.

    Identifier is sent via X-Unique-Identifier-Key / X-Unique-Identifier-Value headers
    per Zoho's spec; the request body carries only the fields to update.
    """
    json_body = _shared.parse_body(body)
    headers = {"X-Unique-Identifier-Key": key, "X-Unique-Identifier-Value": value}
    if upsert:
        headers["X-Upsert"] = "true"
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(BASE, json_body=json_body, headers=headers)
    _shared.emit_object(resp)


@app.command("delete")
def delete_bill(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
):
    """Delete a bill by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{bill_id}")
    _shared.emit_action("bill_id", bill_id, resp)


@app.command("mark-void")
def mark_void(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
):
    """Mark a bill as void (POST /bills/{id}/status/void)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{bill_id}/status/void")
    _shared.emit_action("bill_id", bill_id, resp)


@app.command("mark-open")
def mark_open(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
):
    """Mark a bill as open (POST /bills/{id}/status/open)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{bill_id}/status/open")
    _shared.emit_action("bill_id", bill_id, resp)


@app.command("email")
def email_bill(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    body: str = typer.Option(
        None,
        "--body",
        "-b",
        help="Optional JSON body (to/cc/subject/body); empty body sends Zoho's default.",
    ),
):
    """Email a bill (POST /bills/{id}/email)."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{bill_id}/email", json_body=json_body)
    _shared.emit_action("bill_id", bill_id, resp)


# --- payments sub-app --------------------------------------------------------


@payments_app.command("list")
def list_payments(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List payments applied to a bill."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            f"{BASE}/{bill_id}/payments",
            q,
            "payments",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


@payments_app.command("apply")
def apply_payment(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    body: str = typer.Option(
        ...,
        "--body",
        "-b",
        help=(
            'JSON body wrapping the application rows, e.g. {"bill_payments":'
            '[{"payment_id":"...","amount_applied":100}]}. IDs must be strings.'
        ),
    ),
):
    """Apply existing payments or vendor credits to a bill (POST /bills/{id}/payments).

    The endpoint applies pre-existing payment / vendor-credit records to this
    bill; it does not create new payment records. Use the customer-payments or
    `zb raw POST /vendorpayments` paths for record creation.
    """
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{bill_id}/payments", json_body=json_body)
    _shared.emit_object(resp)


@payments_app.command("delete")
def delete_payment(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    bill_payment_id: str = typer.Argument(..., help="Zoho Books bill_payment_id."),
):
    """Unapply a payment from a bill (DELETE /bills/{id}/payments/{bill_payment_id}).

    Removes only the application row that links this bill to the payment; the
    underlying payment record (vendor-payment / vendor-credit) is untouched and
    can be re-applied or applied to a different bill afterwards.
    """
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{bill_id}/payments/{bill_payment_id}")
    _shared.emit_action("bill_payment_id", bill_payment_id, resp)


# --- comments sub-app --------------------------------------------------------


@comments_app.command("list")
def list_comments(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List recent activity and comments on a bill (read-only)."""
    q: dict[str, str] = {}
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            f"{BASE}/{bill_id}/comments",
            q,
            "comments",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


# --- attachments sub-app -----------------------------------------------------


@attachments_app.command("add")
def attachments_add(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    files: list[Path] = typer.Argument(..., help="One or more files (≤10 MB each)."),
):
    """Attach one or more files to a bill.

    Partial failure is tolerated: each file is uploaded independently and the
    per-file outcome is returned in the result array.
    """
    cfg = config.load()
    results: list[dict] = []
    with ZohoBooksClient(cfg) as client:
        for file in files:
            entry: dict = {"file": str(file)}
            try:
                _uploads.validate(file)
                with file.open("rb") as fh:
                    multipart = {"attachment": (file.name, fh, _uploads.guess_mime(file))}
                    resp = client.post(f"{BASE}/{bill_id}/attachment", files=multipart)
                entry["ok"] = True
                entry["response"] = resp
            except ZohoCLIError as e:
                entry["ok"] = False
                entry["error"] = {
                    "code": e.code,
                    "message": e.message,
                    "details": e.details,
                }
            results.append(entry)
    output.emit_success({"bill_id": bill_id, "results": results})


@attachments_app.command("get")
def attachments_get(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
    output_path: Path = typer.Option(
        ..., "--output", "-o", help="Path to write the downloaded attachment."
    ),
):
    """Download the file attached to a bill."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        body, content_type = client.get_bytes(f"{BASE}/{bill_id}/attachment")
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_bytes(body)
    output.emit_success(
        {
            "bill_id": bill_id,
            "saved_to": str(output_path),
            "size_bytes": len(body),
            "content_type": content_type,
        }
    )


@attachments_app.command("delete")
def attachments_delete(
    bill_id: str = typer.Argument(..., help="Zoho Books bill_id."),
):
    """Delete attachments from a bill."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{bill_id}/attachment")
    output.emit_success({"bill_id": bill_id, "deleted": True, "response": resp})

File: src/commands/bills.rs
//! `zb bills` — CRUD + state actions + email + payments + comments + attachments.

use std::fs;
use std::path::PathBuf;

use clap::{Args, Subcommand};
use serde_json::{Value, json};

use crate::cli::Ctx;
use crate::client::{FileUpload, RequestOptions};
use crate::commands::common::{self, BodyArgs, CustomFieldUpdateArgs, ListArgs};
use crate::errors::{Result, ZohoError};
use crate::shared::Query;
use crate::uploads;

const BASE: &str = "/bills";
const ID: &str = "bill_id";

#[derive(Args, Debug)]
pub struct Cmd {
    #[command(subcommand)]
    pub sub: Sub,
}

#[derive(Subcommand, Debug)]
pub enum Sub {
    List(ListArgs),
    Create(BodyArgs),
    Get(IdArgs),
    Update(UpdateArgs),
    #[command(name = "update-by-custom-field")]
    UpdateByCustomField(CustomFieldUpdateArgs),
    Delete(IdArgs),
    #[command(name = "mark-void")]
    MarkVoid(IdArgs),
    #[command(name = "mark-open")]
    MarkOpen(IdArgs),
    Email(EmailArgs),
    Payments(PaymentsCmd),
    Comments(CommentsCmd),
    Attachments(AttachmentsCmd),
}

#[derive(Args, Debug)]
pub struct IdArgs {
    pub bill_id: String,
}

#[derive(Args, Debug)]
pub struct UpdateArgs {
    pub bill_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct EmailArgs {
    pub bill_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct PaymentsCmd {
    #[command(subcommand)]
    pub sub: PaymentsSub,
}

#[derive(Subcommand, Debug)]
pub enum PaymentsSub {
    /// List payments applied to a bill.
    List(PaymentsListArgs),
    /// Apply existing payments or vendor credits to a bill.
    Apply(PaymentsApplyArgs),
    /// Unapply a payment from a bill.
    Delete(PaymentDeleteArgs),
}

#[derive(Args, Debug)]
pub struct PaymentsListArgs {
    pub bill_id: String,
    #[command(flatten)]
    pub list: ListArgs,
}

#[derive(Args, Debug)]
pub struct PaymentsApplyArgs {
    pub bill_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct PaymentDeleteArgs {
    pub bill_id: String,
    pub bill_payment_id: String,
}

#[derive(Args, Debug)]
pub struct CommentsCmd {
    #[command(subcommand)]
    pub sub: CommentsSub,
}

#[derive(Subcommand, Debug)]
pub enum CommentsSub {
    /// List recent activity and comments on a bill.
    List(CommentsListArgs),
}

#[derive(Args, Debug)]
pub struct CommentsListArgs {
    pub bill_id: String,
    #[command(flatten)]
    pub list: ListArgs,
}

#[derive(Args, Debug)]
pub struct AttachmentsCmd {
    #[command(subcommand)]
    pub sub: AttachmentsSub,
}

#[derive(Subcommand, Debug)]
pub enum AttachmentsSub {
    Add(AttachmentsAddArgs),
    Get(AttachmentsGetArgs),
    Delete(IdArgs),
}

#[derive(Args, Debug)]
pub struct AttachmentsAddArgs {
    pub bill_id: String,
    pub files: Vec<PathBuf>,
}

#[derive(Args, Debug)]
pub struct AttachmentsGetArgs {
    pub bill_id: String,
    #[arg(short = 'o', long)]
    pub output: PathBuf,
}

pub fn run(cmd: Cmd, ctx: &mut Ctx) -> Result<()> {
    match cmd.sub {
        Sub::List(args) => common::list(ctx, BASE, &args, "bills"),
        Sub::Create(args) => common::create(ctx, BASE, &args),
        Sub::Get(args) => common::get(ctx, &format!("{BASE}/{}", args.bill_id)),
        Sub::Update(args) => common::update(ctx, &format!("{BASE}/{}", args.bill_id), &args.body),
        Sub::UpdateByCustomField(args) => common::update_custom(ctx, BASE, &args),
        Sub::Delete(args) => {
            let path = format!("{BASE}/{}", args.bill_id);
            common::delete(ctx, &path, ID, &args.bill_id)
        }
        Sub::MarkVoid(args) => {
            let path = format!("{BASE}/{}/status/void", args.bill_id);
            common::action(ctx, &path, ID, &args.bill_id)
        }
        Sub::MarkOpen(args) => {
            let path = format!("{BASE}/{}/status/open", args.bill_id);
            common::action(ctx, &path, ID, &args.bill_id)
        }
        Sub::Email(args) => {
            let path = format!("{BASE}/{}/email", args.bill_id);
            common::action_with_body(ctx, &path, &args.body, ID, &args.bill_id)
        }
        Sub::Payments(p) => match p.sub {
            PaymentsSub::List(args) => {
                let path = format!("{BASE}/{}/payments", args.bill_id);
                common::list(ctx, &path, &args.list, "payments")
            }
            PaymentsSub::Apply(args) => {
                let path = format!("{BASE}/{}/payments", args.bill_id);
                common::create(ctx, &path, &args.body)
            }
            PaymentsSub::Delete(args) => {
                let path = format!("{BASE}/{}/payments/{}", args.bill_id, args.bill_payment_id);
                common::delete(ctx, &path, "bill_payment_id", &args.bill_payment_id)
            }
        },
        Sub::Comments(c) => match c.sub {
            CommentsSub::List(args) => {
                let path = format!("{BASE}/{}/comments", args.bill_id);
                common::list(ctx, &path, &args.list, "comments")
            }
        },
        Sub::Attachments(a) => match a.sub {
            AttachmentsSub::Add(args) => attachments_add(args, ctx),
            AttachmentsSub::Get(args) => attachments_get(args, ctx),
            AttachmentsSub::Delete(args) => {
                let path = format!("{BASE}/{}/attachment", args.bill_id);
                let resp = ctx.client.delete(&path, &Query::new())?;
                common::emit_success_raw(
                    &json!({
                        "bill_id": args.bill_id,
                        "deleted": true,
                        "response": resp,
                    }),
                    ctx,
                )
            }
        },
    }
}

fn attachments_add(args: AttachmentsAddArgs, ctx: &mut Ctx) -> Result<()> {
    let mut results: Vec<Value> = Vec::new();
    for file in &args.files {
        let mut entry = serde_json::Map::new();
        entry.insert("file".into(), Value::String(file.display().to_string()));
        match upload_one(&args.bill_id, file, ctx) {
            Ok(resp) => {
                entry.insert("ok".into(), Value::Bool(true));
                entry.insert("response".into(), resp);
            }
            // DryRunOk is a sentinel — propagate to short-circuit. Invariant 12 + 14.
            Err(e) if crate::errors::ErrorKind::DryRunOk == e.kind => return Err(e),
            Err(e) => {
                entry.insert("ok".into(), Value::Bool(false));
                entry.insert(
                    "error".into(),
                    json!({
                        "code": e.code(),
                        "message": e.message,
                        "details": e.details.clone().unwrap_or_else(|| json!({})),
                    }),
                );
            }
        }
        results.push(Value::Object(entry));
    }
    common::emit_success_raw(&json!({"bill_id": args.bill_id, "results": results}), ctx)
}

fn upload_one(bill_id: &str, file: &std::path::Path, ctx: &mut Ctx) -> Result<Value> {
    uploads::validate(file)?;
    let opts = RequestOptions {
        files: vec![FileUpload {
            field: "attachment".into(),
            path: file.to_path_buf(),
        }],
        ..RequestOptions::default()
    };
    ctx.client
        .post(&format!("{BASE}/{bill_id}/attachment"), opts)
}

fn attachments_get(args: AttachmentsGetArgs, ctx: &mut Ctx) -> Result<()> {
    let (bytes, content_type) = ctx.client.get_bytes(
        &format!("{BASE}/{}/attachment", args.bill_id),
        &Query::new(),
    )?;
    if let Some(parent) = args.output.parent() {
        fs::create_dir_all(parent).map_err(ZohoError::from)?;
    }
    fs::write(&args.output, &bytes).map_err(ZohoError::from)?;
    common::emit_success_raw(
        &json!({
            "bill_id": args.bill_id,
            "saved_to": args.output.display().to_string(),
            "size_bytes": bytes.len(),
            "content_type": content_type,
        }),
        ctx,
    )
}
